src/suites/cts/validation.spec.ts

```diff
@@ -0,0 +1,673 @@
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: 'sampler',
+      },
+    ],
+  });
+
+  const mismatchedResources = [
+    {
+      buffer: t.uniformBuffer,
+    },
+    {
+      buffer: t.storageBuffer,
+    },
+    {
+      buffer: t.getErrorBuffer(),
+    },
+    t.sampledTexture.createView(),
+  ];
+
+  await t.expectBindGroupValidationErrors(bindGroupLayout, mismatchedResources);
+});
+
+g.test('validation/texture view binding must contain exactly one texture view', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: 'sampled-texture',
+      },
+    ],
+  });
+
+  const mismatchedResources = [
+    {
+      buffer: t.uniformBuffer,
+    },
+    {
+      buffer: t.storageBuffer,
+    },
+    {
+      buffer: t.getErrorBuffer(),
+    },
+    t.sampler,
+  ];
+
+  await t.expectBindGroupValidationErrors(bindGroupLayout, mismatchedResources);
+});
+
+g.test('validation/texture must have correct usage', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: 'sampled-texture',
+      },
+    ],
+  });
+
+  const outputTexture = t.device.createTexture({
+    size: { width: 16, height: 16, depth: 1 },
+    format: 'r8unorm',
+    usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+  });
+
+  t.device.createBindGroup({
+    bindings: [
+      {
+        binding: 0,
+        resource: outputTexture.createView(),
+      },
+    ],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/texture must have correct component type', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: 'sampled-texture',
+      },
+    ],
+  });
+
+  const uintTexture = t.device.createTexture({
+    size: { width: 16, height: 16, depth: 1 },
+    format: 'rgba8uint',
+    usage: GPUTextureUsage.SAMPLED,
+  });
+
+  t.device.createBindGroup({
+    bindings: [
+      {
+        binding: 0,
+        resource: uintTexture.createView(),
+      },
+    ],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/buffer offset for bind groups match', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+    ],
+  });
+
+  const invalidBufferConstraints = [
+    {
+      offset: 1,
+      size: 256,
+    },
+    {
+      offset: 128,
+      size: 256,
+    },
+    {
+      offset: 255,
+      size: 256,
+    },
```

Comment:
This one should be easy to parameterize, something like this: https://github.com/gpuweb/cts/blob/master/src/suites/cts/examples.spec.ts#L60